src/riak_kv_ts_api.erl

```diff
@@ -97,6 +104,93 @@ api_calls() ->
+        {bad_linkfun_bkey, {B, K}} ->
+            {error, SvcMod:make_table_create_fail_resp(Table,
+                                                       flat_format(
+                                                         "Malformed bucket/key for anon link fun in bucket type properties: ~p/~p\n", [B, K]))};
+        {bad_chash_keyfun, {M, F}} ->
+            {error, SvcMod:make_table_create_fail_resp(Table,
+                                                       flat_format(
+                                                         "Invalid chash mod or fun in bucket type properties: ~p:~p\n", [M, F]))};
+        Props2 ->
+            create_table1(SvcMod, Table, Props2, DDL2, DDLRecCap, ?TABLE_ACTIVATE_WAIT)
+    end.
+
+create_table1(SvcMod, Table, Props, DDL, DDLRecCap, Seconds) ->
+    case riak_core_bucket_type:create(Table, Props) of
+        ok -> wait_until_active_and_supported(SvcMod, Table, DDL, DDLRecCap, Seconds);
+        {error, Reason} -> {error, SvcMod:make_table_create_fail_resp(Table, Reason)}
+    end.
+
+wait_until_supported(SvcMod, Table, _DDL, _DDLRecCap, _Seconds=0) ->
+    {error, SvcMod:make_table_activate_error_timeout_resp(Table)};
+wait_until_supported(SvcMod, Table, DDL, DDLRecCap, Seconds) ->
+    case get_active_peer_nodes() of
+        Nodes when is_list(Nodes) ->
+            wait_until_supported1(Nodes, SvcMod, Table, DDL, DDLRecCap, Seconds);
+        {error, _Reason} ->
+            timer:sleep(1000),
+            wait_until_supported(SvcMod, Table, DDL, DDLRecCap, Seconds - 1)
+    end.
+
+wait_until_supported1(Nodes, SvcMod, Table, DDL, DDLRecCap, Seconds) ->
+    case get_remote_is_table_active_and_supported(Nodes, Table, DDLRecCap) of
+        true -> ok;
+        _ ->
+            timer:sleep(1000),
+            lager:info("Waiting for table ~ts to be compiled on all active nodes", [Table]),
+            wait_until_supported1(Nodes, SvcMod, Table, DDL, DDLRecCap, Seconds - 1)
+    end.
+
+get_remote_is_table_active_and_supported(Nodes, Table, DDLRecCap) ->
+    multi_is_all_true(
+      rpc:multicall(Nodes, riak_kv_ts_util, is_table_supported,
+                    [DDLRecCap, Table])).
+
+multi_is_all_true({_Results, BadNodes}) when BadNodes =/= [] ->
+    false;
+multi_is_all_true({Results, _BadNodes}) ->
+    lists:all(fun(E) -> E == true end, Results).
+
+get_active_peer_nodes() ->
+    case riak_core_ring_manager:get_my_ring() of
+        {ok, Ring} -> riak_core_ring:active_members(Ring);
+        _ -> {error, retry}
+    end.
+
+wait_until_active_and_supported(SvcMod, Table, _DDL, _DDLRecCap, _Seconds=0) ->
+    {error, SvcMod:make_table_activate_error_timeout_resp(Table)};
+wait_until_active_and_supported(SvcMod, Table, DDL, DDLRecCap, Seconds) ->
+    case riak_core_bucket_type:activate(Table) of
+        ok -> wait_until_supported(SvcMod, Table, DDL, DDLRecCap, Seconds);
+        {error, not_ready} ->
+            timer:sleep(1000),
```

Comment:
I believe the expectation here is that the table gets activated quickly -- but not immediately -- under normal circumstances. It is likely, in other words, that the first time `riak_core_bucket_type:activate(Table)` will return `{error, not_ready}` and a second attempt made in 10 msec would succeed. Introducing a small delay on the initial iteration would, then, give you a fair chance of going through it in (much) less than a second.